src/evaluators/JSXElement.js

```diff
@@ -221,90 +223,153 @@ function evaluateJSXAttributes(
-                } else {
-                  attributes.set(key, spreadPropValue);
-                }
-              }
-            }
+          for (let [spreadPropKey] of spreadValue.properties) {
+            setProp(spreadPropKey, Get(realm, spreadValue, spreadPropKey), false);
           }
         } else {
-          throw new FatalError("ObjectValues are the only supported value for JSX Spread Attributes");
+          spreadAttributesCount++;
+          containsAbstractSpreadAttribute = true;
+          // check to see if this object has an abstractHint
+          invariant(spreadValue instanceof AbstractValue);
+          let reactHint = realm.react.abstractHints.get(spreadValue);
+
+          if (reactHint === "HAS_NO_KEY_OR_REF") {
+            spreadAttributesWithInitialPropsHintCount++;
+            // as we know initial props can't have a "props" or "key", we don't have
+            // to mark them as abstract
+          } else {
+            // as the spread might contain "key" and "ref" we have to mark both
+            // as being abstract too, as we don't know if they exist
+            containsAbstractKey = true;
+            containsAbstractRef = true;
+          }
+          // we push the props up to this point into the abstract props args. we also
+          // push the abstract spread object and then we create a fresh props object
+          abstractPropsArgs.push(props, spreadValue);
+          props = Create.ObjectCreate(realm, realm.intrinsics.ObjectPrototype);
         }
         break;
       default:
-        invariant(false, `Unknown JSX attribute type:: ${astAttribute.type}`);
+        invariant(false, `Unknown JSX attribute type: ${astAttribute.type}`);
     }
   }
-  return {
-    attributes,
-    children,
-  };
-}
-
-function createReactProps(
-  realm: Realm,
-  type: Value,
-  attributes: Map<string, Value>,
-  children,
-  env: LexicalEnvironment
-): ObjectValue {
-  let obj = Create.ObjectCreate(realm, realm.intrinsics.ObjectPrototype);
-  for (let [key, value] of attributes) {
-    if (typeof key === "string") {
-      if (RESERVED_PROPS.hasOwnProperty(key)) {
-        continue;
+  if (containsAbstractSpreadAttribute) {
+    // if we haven't assigned any attributes and we are dealing with a single
+    // spread attribute, we can just make the spread object the props
+    if (attributesAssigned === 0) {
+      props = abstractPropsArgs[1];
+    } else {
+      // we create an abstract Object.assign() to deal with the fact that we don't what
+      // the props are because they contain abstract spread attributes that we can't
+      // evaluate ahead of time
+      let types = new TypesDomain(FunctionValue);
+      let values = new ValuesDomain();
+      let emptyObject = Create.ObjectCreate(realm, realm.intrinsics.ObjectPrototype);
+      invariant(realm.generator);
+      props = realm.generator.derive(types, values, [emptyObject, ...abstractPropsArgs], _args => {
```

Comment:
Not sure this is right. `abstractPropsArgs` are only created from spreads.
But you may have something like `<div {...props1} something="something" {...props2} />`.
I think this code will do `Object.assign({}, props1, props2)` and lose `something="something"`.

Reply:
Hmm, I see now you're pushing props *up to that point* to the array.

Still I'd like to see more coverage here. A bunch of different cases with one spread, many spreads, spread-before-normal-args, spread-after-normal-args, spread-between-args, etc. Make sure `key` and `ref` end up being what we expect in all cases, even with complex overrides.

Reply:
I'll add more tests for these cases.